python/ray/serve/http_proxy.py

```diff
@@ -1,23 +1,82 @@
 import asyncio
 import socket
-from typing import List
+from typing import List, Dict, Tuple
 
 import uvicorn
 import starlette.responses
+import starlette.routing
 
 import ray
 from ray.exceptions import RayTaskError
+from ray.serve.common import EndpointTag
 from ray.serve.constants import LongPollKey
 from ray.util import metrics
 from ray.serve.utils import _get_logger
 from ray.serve.http_util import Response, build_starlette_request
 from ray.serve.long_poll import LongPollAsyncClient
-from ray.serve.router import Router
 from ray.serve.handle import DEFAULT
 
 logger = _get_logger()
 
 
+class ServeStarletteEndpoint:
+    """Wraps the given Serve endpoint in a Starlette endpoint.
+
+    Implements the ASGI protocol.  Constructs a Starlette endpoint for use by
+    a Starlette app or Starlette Router which calls the given Serve endpoint
+    using the given Serve client.
+
+    Usage:
+        route = starlette.routing.Route(
+                "/api",
+                ServeStarletteEndpoint(self.client, endpoint_tag),
+                methods=methods)
+        app = starlette.applications.Starlette(routes=[route])
+    """
+
+    def __init__(self, client, endpoint_tag: EndpointTag):
+        self.client = client
+        self.endpoint_tag = endpoint_tag
+        self.handle = None
+
+    async def __call__(self, scope, receive, send):
+        http_body_bytes = await self.receive_http_body(scope, receive, send)
+
+        headers = {k.decode(): v.decode() for k, v in scope["headers"]}
+        if self.handle is None:
+            self.handle = self.client.get_handle(self.endpoint_tag, sync=False)
+        self.handle = self.handle.options(
+            method_name=headers.get("X-SERVE-CALL-METHOD".lower(),
+                                    DEFAULT.VALUE),
+            shard_key=headers.get("X-SERVE-SHARD-KEY".lower(), DEFAULT.VALUE),
+            http_method=scope["method"].upper(),
+            http_headers=headers)
```

Comment:
nit, but we don't need to store the modified handle because it's only ever used once. We could just do `self.handle.options(...).remote(request)` below.

Reply:
I think that would make sense if `handle.options()` just modified the existing ServeHandle, but from the implementation (`return self.__class__(self.router, self.endpoint_name, new_options)`)
it looks like `handle.options()` actually constructs and returns a new ServeHandle.  So don't we actually need to store the modified handle here?